### ISSUE TITLE
Missing `SectionType` for parsing parts of data file (BUG?): `RUNSPEC` and `EDIT` 

### ISSUE THREAD

**Issue by asnyv** [OPM/opm-common#3855]
Functionality for parsing parts of a .DATA-file was introduced in #2527, but for a for me unknown reason the sections `RUNSPEC` (required) and `EDIT` (optional) were left out. Would be nice to get these included.

**Comment by asnyv** [comment]
Did some research/testing

- Seems like the reason why `RUNSPEC` was left out is that it is always included in the parsing due to some keywords in other sections requiring RUNSPEC to be properly parsed (e.g. `EQLDIMS`). This is though maybe not a reason for why you couldn't read _only_ `RUNSPEC`? So I think this should still be included. Also it would be very nice if we could drop the keywords from `RUNSPEC` section in the output even though it has to be read in the initial parsing, as one could then use the functionality to get individual sections of the DATA-file without the need of postprocessing. As far as I know there isn't a direct way to split a parsed deck into sections, except searching through the deck records for section keywords.

- The `EDIT` section is likely forgotten due to it being optional and not present in the `DATA` file that was selected as test case.

- The optional `EDIT` and `REGIONS` sections make the [check_sections_keyword](https://github.com/OPM/opm-common/blob/8239c90731c27814f4a23241100fb97be8a3a38b/src/opm/input/eclipse/Parser/Parser.cpp#L570) validation today a bit buggy as `REGIONS` is required and `EDIT` is missing (+ `RUNSPEC` if that is to be introduced). Not sure how the optional keywords should be handled in such a case. Don't know very well how the parser is built up, but an idea could be that it is required that the section(s) you specify that you want to parse have to be present in the root file (+ `RUNSPEC`). Then you might end up reading too much if some sections are hidden in included files, but they could maybe later be dropped in the output (similar to what I suggested to do with the `RUNSPEC` when that is not requested)?

**Comment by bska** [comment]
@asnyv: Just a quick update here: A candidate solution (PR #3863) has been merged into the master development sources.  If you're comfortable build the simulator from source code, you can have access to that solution now and use it for testing purposes.  In that case, any feedback is greatly welcome.  Otherwise, the feature will be available in the next release of the OPM Flow simulator&ndash;designation 2024.04&ndash;which is tentatively scheduled for late April 2024.

**Comment by asnyv** [comment]
@bska @tskille :
Thanks! 
I did a quick test (through the python api). Seems like there is a small bug:
From my tests, the identifying keyword for the sections after `RUNSPEC` (`GRID`, `EDIT` and etc) becomes incorrect unless you have also requested the previous section. Seems like the section keyword returned is always the first the parser needed to check and not the first in the section requested. 

Example. Let the following be the start of X.DATA:
```
RUNSPEC
KEYWORD1
KEYWORD2

GRID
KEYWORD3
KEYWORD4

EDIT
KEYWORD5
KEYWORD6

PROPS
KEYWORD7
```
By parsing with 
`opm.io.parser.Parser().parse("X.DATA", opm.io.parser.ParseContext(some_context), [opm.io.eclSectionType.GRID]`
I would expect to get
`GRID, KEYWORD3, KEYWORD4`
but instead I get
`RUNSPEC, KEYWORD3, KEYWORD4`

Likewise, if I request multiple sections the second requested starts with the keyword of the first section after the first requested section. E.g. when parsing:
`opm.io.parser.Parser().parse("X.DATA", opm.io.parser.ParseContext(some_context), [opm.io.eclSectionType.RUNSPEC, opm.io.eclSectionType.PROPS]`
I would expect to get
`RUNSPEC, KEYWORD1, KEYWORD2, PROPS, KEYWORD7`
but instead I get
`RUNSPEC, KEYWORD1, KEYWORD2, GRID, KEYWORD7`
As GRID is the first section after RUNSPEC.

So to sum up, all sections except `RUNSPEC` becomes wrong unless the previous section is also requested.

My guess is that this is an indexing error when removing keywords after parsing, and that one should have started removing keywords one index earlier and ended one index earlier.

**Comment by bska** [comment]
@asnyv : I'd like to give another update here.  The recent PR #3896 supposedly fixes this problem for real.  Would you mind testing a recent build to see if the issue has been resolved?

**Comment by asnyv** [comment]
@bska: I already tested the PR and commented there. I can also test it again now that it has been merged if you want me to.

**Comment by bska** [comment]
> I already tested the PR and commented there.

Ah, right.  I must have missed that.

> I can also test it again now that it has been merged if you want me to.

That's up to you, I think.  If you're satisfied that the issue has been resolved then we can simply close this now.  Otherwise, it doesn't hurt to have one more piece of confirmation.

**Comment by bska** [comment]
> I think it should be fine to close the issue

Very good.  I'll close the issue then.
